File: Difficulty: Easy/Reverse Words/reverse-words.java
//{ Driver Code Starts
import java.util.*;
import java.lang.*;
import java.io.*;
class GFG {
    public static void main(String[] args) {

        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();
        while (t > 0) {
            String s = sc.next();
            Solution obj = new Solution();
            System.out.println(obj.reverseWords(s));
            t--;
        }
    }
}

// } Driver Code Ends




class Solution 
{
    //Function to reverse words in a given string.
    String reverseWords(String S)
    {
        // code here 
        char str [] = S.toCharArray();
        int start = 0;
        for(int end = 0; end<str.length; end ++){
            if(str[end]=='.'){
                rev(str, start, end-1);
                start = end +1;
            }
        }
        rev(str, start, str.length-1);
        rev(str, 0, str.length-1);
        return  new String(str);
    }
    static void rev(char str[], int start, int end){
        char temp;
        while(start<end){
            temp = str[start];
            str[start] = str[end];
            str[end] = temp;
            start ++;
            end --;
        }
    }
}

File: Difficulty: Easy/Reverse Words/reverse-words.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends


class Solution {
  public:
    // Function to reverse words in a given string.
    string reverseWords(string str) {
        stack<string>s;
        int n=str.size();
        int i=0,j=0;
        string ans="";
        while(i<n && j<n)
        {
            if(j==n-1)
            {
                string x= str.substr(i,(j-i)+1);
                s.push(x);
                i=j+1; 
            }
            else if(str[j]=='.')
            {
                string x= str.substr(i,(j-i));
                s.push(x);
                i=j+1;
            }
            j++;
        }
        while(!s.empty())
        {
            string x= s.top();
            ans+=x;
            s.pop();
            if(s.size()>0)
            ans.push_back('.');
        }
        return ans;
    }
};

//{ Driver Code Starts.
int main() {
    int t;
    cin >> t;
    while (t--) {
        string s;
        cin >> s;
        Solution obj;
        cout << obj.reverseWords(s) << endl;
    }
}
// } Driver Code Ends